### DISCUSSION TITLE
hipsycl-hcf-tool

### DISCUSSION THREAD

**Discussion by BeauJoh** [AdaptiveCpp/AdaptiveCpp#760] [General]
Hey!

I was wondering if I could extract the compiled SYCL kernel binaries (from OpenMP, CUDA and HIP backends) to run as standalone kernel with all the tedious backend-specific garb.

Currently, I compile the binary with:

```
HIPSYCL_HCF_DUMP_DIRECTORY=/workspace/dump /opt/hipSYCL/bin/syclcc --hipsycl-targets='hip:gfx906' ./sycl-matmul.cpp
```

Unfortunately, the dump directory is empty (which I had hoped would be populated with HCF content. And running the HCF tool on the binary doesn't give any goodies:

```
/opt/hipSYCL/bin/hipsycl-hcf-tool a.out  -p root

terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::erase: __pos (which is 18446744073709551615) > this->s
ize() (which is 1111984)
Aborted (core dumped)
```

Any thoughts/suggestions would be appreciated.

Cheers!

**Comment by illuhad** [comment]
Hi!

In principle your approach is correct: Setting this environment variable should populate the directory with `*.hcf` files, which can then be inspected or modified using `hipsycl-hcf-tool`.
Note that hipSYCL does not create directories, so the dump directory must already exist.

Also, HCF only plays a role if hipSYCL controls the multipass compilation and embedding of compiled device binaries into the host application. We currently distinguish integrated multipass compilation (clang or some other third-party compiler does multipass compilation and embedding of kernels) flows and explicit multipass (hipSYCL controls everything) compilation flows. See https://github.com/illuhad/hipSYCL/blob/develop/doc/compilation.md for details.

So, HCF is only used in explicit multipass scenarios, since we have no control over the kernel format otherwise.
Currently, explicit multipass is only supported by the CUDA and SPIR-V backends. On HIP, it is implementable but implementation priorities mean that it is not there yet (but of course, a PR is welcome ;) ).
On OpenMP, this will probably never work because OpenMP code is just host code, and as such not separated from the host application.

In integrated multipass scenarios such as HIP in your example, you might still be able to access the kernels using clang-specific flags during compilation. (E.g. `-emit-llvm` and/or `-S`). With this, you should be able to get the generated amdgcn code for HIP.

**EDIT**: I've just noticed that you are using the environment variable with `syclcc`. This won't work; HCF dump functionality is a runtime thing, so you need to set this environment variable when you run the compiled application. When the hipSYCL runtime starts, it will then dump all HCF objects that it knows about. This can allow you to access compiled SYCL kernels even if you don't have access to the source code (e.g. from external shared libraries).
For compile-time use cases, you have two options:
* Backend specific clang flags for integrated multipass, as mentioned above
* For explicit multipass, you can use `syclcc --hipsycl-save-temps` which will tell hipSYCL to retain temporary files (including compiled device binaries that it creates during compilation. It will print the directory where you can find them.
